Sources/Data/Attributes.swift

```diff
@@ -0,0 +1,299 @@
+                }
+                
+                // now try probing for an embedded dict.
+                if let dictionary = try? container.nestedContainer(keyedBy: DynamicCodingKeys.self, forKey: key) {
+                    assembledHash[keyString] = try fromKeyedDecoder(dictionary)
+                    return
+                }
+                
+                // now try probing for an array.
+                if var array = try? container.nestedUnkeyedContainer(forKey: key) {
+                    var collection: [Any] = []
+                    while(!array.isAtEnd) {
+                        if let value = try? array.decode(Bool.self) {
+                            collection.append(value)
+                            continue
+                        }
+                        
+                        if let value = try? array.decode(Int.self) {
+                            collection.append(value)
+                            continue
+                         }
+                        
+                        if let value = try? array.decode(String.self) {
+                            collection.append(value)
+                            continue
+                        }
+                        
+                        throw DecodingError.dataCorruptedError(in: array, debugDescription: "Expected one of Int, String, Double, Boolean. Rover attributes arrays may only contain those primitive types.")
+                    }
+                    
+                    assembledHash[keyString] = collection
+                    return
+                }
+                
+                throw DecodingError.dataCorruptedError(forKey: key, in: container, debugDescription: "Expected one of Int, String, Double, Boolean, or an Array thereof.")
+            }
+            
+            return assembledHash
+        }
+        
+        let container = try decoder.container(keyedBy: DynamicCodingKeys.self)
+        
+        self.rawValue = try fromKeyedDecoder(container)
+    }
+    
+    public func encode(to encoder: Encoder) throws {
+        /// nested function for recursing through the dictionary and populating the Encoder with it, doing the necessary type coercions on the way.
+        func encodeToContainer(dictionary: [String: Any], container: inout KeyedEncodingContainer<Attributes.DynamicCodingKeys>) throws {
+            try dictionary.forEach { (codingKey, value) in
+                let key = DynamicCodingKeys(stringValue: codingKey)
+                switch value {
+                case let value as Int:
+                    try container.encode(value, forKey: key)
+                case let value as Bool:
+                    try container.encode(value, forKey: key)
+                case let value as String:
+                    try container.encode(value, forKey: key)
+                case let value as Double:
+                    try container.encode(value, forKey: key)
+                case let array as [Any]:
```

Comment:
Similar to the comment made in decoder, you can replace this with simple casting cases. E.g.:
```swift
case let value as [String]:
    ...
case let value as [Double]:
    ...
```

Reply:
Likewise.